server/lib/src/idm/oauth2.rs

```diff
@@ -236,6 +238,66 @@ pub struct AuthorisePermitSuccess {
     pub state: String,
     // The exchange code as a String
     pub code: String,
+    /// The format the response should be returned to the application in.
+    pub response_mode: ResponseMode,
+}
+
+impl AuthorisePermitSuccess {
+    /// Builds a redirect URI to go back to the application when permission was
+    /// granted.
+    pub fn build_redirect_uri(&self) -> Url {
+        let mut redirect_uri = self.redirect_uri.clone();
+
+        // Always clear query and fragment, regardless of the response mode
+        redirect_uri.set_query(None);
+        redirect_uri.set_fragment(None);
+
+        // We can't set query pairs on fragments, only query.
+        let encoded = url::form_urlencoded::Serializer::new(String::new())
+            .append_pair("state", &self.state)
+            .append_pair("code", &self.code)
+            .finish();
+
+        match self.response_mode {
+            ResponseMode::Query => redirect_uri.set_query(Some(&encoded)),
+            ResponseMode::Fragment => redirect_uri.set_fragment(Some(&encoded)),
+        }
+
+        redirect_uri
+    }
+}
+
+#[derive(Debug)]
+pub struct AuthoriseReject {
+    // Where the RS wants us to go back to.
```

Comment:
Client, not RS. It's a language mistake from the past.

Reply:
Fixed.